### ISSUE TITLE
Question: Does the ZFS kernel modul benefit from compiler option -march=znver2 or znver3?

### ISSUE THREAD

**Issue by mabod** [openzfs/zfs#13202]
I do not know where to ask that question so I put it here as a bug report. Sorry for that.

With the graysky2 kernel compiler patch (https://github.com/graysky2/kernel_compiler_patch) it is possible to compile the  linux kernel with compiler optimizations for Ryzen processors for example (-march=znver2 or -march=znver3).

This patch is already included in linux-xanmod and linux-zen. And if the kernel is compiled with -march=znver2 the zfs module is also compiled with that option. I tested that.

Does that make any difference? Is that good or bad for the zfs performance or stability? I was wondering if encryption or compression would benefit from that.

**Comment by rincebrain** [comment]
There's a Discussions tab that this would fit in.

"Sort of, maybe?"

Some of the benefits you get from using -march=[whatever] are being able to use various platform-specific extensions like AVX2, and you absolutely cannot do that without setup on random functions in the kernel unless you want to burn the world down, so the kernel build flags have a million arch-specific -mno-dont-do-that-ever.

Some of it is taking advantage of knowledge of what will and won't fit in cache and the like, and it could presumably benefit from that in general.

But most of the major perf gains in encryption would commonly be from leveraging various SIMD-accelerated implementations, which we already have a few of, but more could always be welcome if they're a win on your platform. (Of course, below a certain threshold of improvement, staying with purely simple integer math can be a major win, so maybe sometimes...but the speedups can be very dramatic in encryption with hardware acceleration, so probably not on Zen in particular.)

As far as compression, LZ4's authors have said they haven't found a good way to make it use any accelerated instructions at the moment; gzip has some SIMD-accelerated implementations but since we're just using the kernel's builtin zlibs right now AFAIK neither FBSD nor Linux's do that; zstd has BMI2 knobs that it'll use if they're available at runtime (which is fine, since they don't require worrying about clobbering everyone else when you do it), and SSE2 knobs that we currently don't use and would require minor code changes to use (which might or might not be a win...)

[My answer here has overlapping further information](https://github.com/openzfs/zfs/discussions/12470#discussioncomment-2271332).

**Comment by mabod** [comment]
How do I have to understand your answer? I translate this into: march=znver2 has no benefit and it does not hurt either.

**Comment by rincebrain** [comment]
tl;dr "a lot of the performance improvements from using -march come from using instructions you can't freely use in the kernel; it still might help, but not, I would think, earth-shatteringly in most cases"

**Comment by ryao** [comment]
@mabod ZFS does not benefit from `-march` since it spends most of its time waiting on interrupts and the compute heavy portions use optimized assembly.

That said, it would not hurt to use it either. That is what you already heard from @rincebrain though.

I am closing this since there is nothing to fix here.
